genesis/maker.go

```diff
@@ -61,20 +73,252 @@ func NewGenesisValidator(amount int64, name string, unbondToAddress []byte,
+// NOTE: [ben] is introduced as technical debt to preserve the signature
+// of GenerateKnown but in order to introduce the timestamp gradually
+// This will be deprecated in v0.17
+func generateKnownWithTime(chainName, accountsPathCSV, validatorsPathCSV string,
+	genesisTime time.Time) (string, error) {
+	var genDoc *GenesisDoc
+
+	// TODO [eb] eliminate reading priv_val ... [zr] where?
+	if accountsPathCSV == "" || validatorsPathCSV == "" {
+		return "", fmt.Errorf("both accounts.csv and validators.csv is required")
+	}
+
+	pubkeys, amts, names, perms, setbits, err := parseCsv(validatorsPathCSV)
+	if err != nil {
+		return "", err
+	}
+
+	pubkeysA, amtsA, namesA, permsA, setbitsA, err := parseCsv(accountsPathCSV)
+	if err != nil {
+		return "", err
+	}
+
+	genDoc = newGenDoc(chainName, genesisTime, len(pubkeys), len(pubkeysA))
+	for i, pk := range pubkeys {
+		genDocAddValidator(genDoc, pk, amts[i], names[i], perms[i], setbits[i], i)
+	}
+	for i, pk := range pubkeysA {
+		genDocAddAccount(genDoc, pk, amtsA[i], namesA[i], permsA[i], setbitsA[i], i)
+	}
+
+	buf, buf2, n := new(bytes.Buffer), new(bytes.Buffer), new(int)
+	wire.WriteJSON(genDoc, buf, n, &err)
+	if err != nil {
+		return "", err
+	}
+	if err := json.Indent(buf2, buf.Bytes(), "", "\t"); err != nil {
+		return "", err
+	}
+
+	return buf2.String(), nil
+}
+
+//-----------------------------------------------------------------------------
+// gendoc convenience functions
+
+func newGenDoc(chainName string, genesisTime time.Time, nVal, nAcc int) *GenesisDoc {
+	genDoc := GenesisDoc{
+		ChainName:   chainName,
+		GenesisTime: genesisTime,
+	}
+	genDoc.Accounts = make([]Account, nAcc)
+	genDoc.Validators = make([]Validator, nVal)
+	return &genDoc
+}
+
+func genDocAddAccount(genDoc *GenesisDoc, pubKey crypto.PubKeyEd25519, amt uint64, name string,
+	perm, setbit ptypes.PermFlag, index int) {
+	addr, _ := acm.AddressFromBytes(pubKey.Address())
+	acc := Account{
+		BasicAccount: BasicAccount{
+			Address: addr,
+			Amount:  amt,
+		},
+		Name: name,
+		Permissions: ptypes.AccountPermissions{
+			Base: ptypes.BasePermissions{
+				Perms:  perm,
+				SetBit: setbit,
+			},
+		},
+	}
+	if index < 0 {
+		genDoc.Accounts = append(genDoc.Accounts, acc)
+	} else {
+		genDoc.Accounts[index] = acc
+	}
+}
+
+func genDocAddValidator(genDoc *GenesisDoc, pubKey crypto.PubKeyEd25519, amt uint64, name string,
+	perm, setbit ptypes.PermFlag, index int) {
+	addr, _ := acm.AddressFromBytes(pubKey.Address())
+	genDoc.Validators[index] = Validator{
+		BasicAccount: BasicAccount{
+			Address:   acm.MustAddressFromBytes(pubKey.Address()),
+			PublicKey: acm.PublicKeyFromPubKey(pubKey.Wrap()),
+			Amount:    amt,
+		},
+		Name: name,
+		UnbondTo: []BasicAccount{
+			{
+				Address: addr,
+				Amount:  amt,
+			},
+		},
+	}
+	// [zr] why no index < 0 like in genDocAddAccount?
+}
+
+//-----------------------------------------------------------------------------
+// util functions
```

Comment:
We should review the below functions as I'm pretty sure they're redundant with other code or not needed.

Reply:
oops yeah that should be deleted, found its way back in during rebase.